packager/packager.cc

```diff
@@ -1546,9 +1546,83 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
     return info;
 }
 
+void validateLayering(const core::Context &ctx, const Import &i) {
+    if (i.type == ImportType::Test) {
+        return;
+    }
+
+    const auto &packageDB = ctx.state.packageDB();
+    ENFORCE(packageDB.getPackageInfo(i.name.mangledName).exists())
+    ENFORCE(packageDB.getPackageForFile(ctx, ctx.file).exists())
+    auto &thisPkg = PackageInfoImpl::from(packageDB.getPackageForFile(ctx, ctx.file));
+    auto &otherPkg = PackageInfoImpl::from(packageDB.getPackageInfo(i.name.mangledName));
+
+    if (!thisPkg.strictDependenciesLevel.has_value() || !otherPkg.strictDependenciesLevel.has_value() ||
+        !thisPkg.layer.has_value() || !otherPkg.layer.has_value()) {
+        return;
+    }
+
+    if (thisPkg.strictDependenciesLevel.value().first == core::packages::StrictDependenciesLevel::False) {
+        return;
+    }
+
+    auto possibleLayers = packageDB.layers();
+    auto pkgLayer = thisPkg.layer.value().first;
+    auto otherPkgLayer = otherPkg.layer.value().first;
+    auto pkgIndex = packageDB.layerIndex(pkgLayer);
+    auto otherPkgIndex = packageDB.layerIndex(otherPkgLayer);
+
+    if (pkgIndex < otherPkgIndex) {
+        if (auto e = ctx.beginError(i.name.loc, core::errors::Packager::LayeringViolation)) {
+            e.setHeader("`{}` is at layer `{}`, so it can not import package `{}`, which is at layer `{}`",
+                        thisPkg.name.toString(ctx), pkgLayer.toString(ctx), otherPkg.name.toString(ctx),
+                        otherPkgLayer.toString(ctx));
```

Comment:
```suggestion
                        thisPkg.show(ctx), pkgLayer.show(ctx), otherPkg.show(ctx),
                        otherPkgLayer.show(ctx));
```

`NameRef::toString` includes things that are not meant to show up in error messages.

For layers, which are always UTF8 names, this doesn't matter, but we should still use show.

`thisPkg.name.toString(gs)` is the same as `thisPkg.show(gs)` so we can at least make it shorter with no loss. But for the same reason as before, we shouldn't get in the habit of using `toString`-named methods in error messages. Those methods should always be called `.show`. I'm a little sad that we even have a `toString` method on the `PackageName` class—it should probably be called `show`.